topdown/eval.go

```diff
@@ -2841,6 +2856,12 @@ func (e evalVirtualPartial) reduce(rule *ast.Rule, b *bindings, result *ast.Term
 		//          |    leafKey
 		//          objPath
 		fullPath := rule.Ref()
+
+		collisionPath := fullPath[e.pos+1:]
+		if hasCollisions(collisionPath, visitedRefs, b) {
```

Comment:
Alternatively, I wonder if sorting the rules by absolute ref length, higher to lower, would also be a solution 🤔 .